text/3502-cargo-script.md

```diff
@@ -0,0 +1,1229 @@
+- `./file` with `#!/usr/bin/env bash` to make standalone executables
+- `bash -c <expr>` to try out an idea right now
+- Common configuration with rc files, `--rcfile <path>`
+
+Python
+- `python` to get an interactive way of entering code
+- `python -i ...` to make other ways or running interactive
+- `python <file>` will run the file
+- `./file` with `#!/usr/bin/env python` to make standalone executables
+- `python -c <expr>` to try out an idea right now
+- Can run any file in a project (they can have their own "main") to do whitebox exploratory programming and not just blackblox
+
+Go
+- [`gorun`](https://github.com/erning/gorun/) attempts to bring that experience to a compiled language, go in this case
+  - `gorun <file>` to build and run a file
+  - Implicit garbage collection for build cache
+  - Project metadata is specified in HEREDOCs in regular code comments
+
+Perl
+- [Re-interprets the `#!`](https://stackoverflow.com/questions/38059830/how-does-perl-avoid-shebang-loops)
+
+Ruby
+- [`bundler/inline`](https://bundler.io/guides/bundler_in_a_single_file_ruby_script.html)
+  - Uses a code-block to define dependencies, making them available for use
+
+Cross-language
+- [`scriptisto`](https://github.com/igor-petruk/scriptisto)
+  - Supports any compiled language
+  - Comment-directives give build commands
+- [nix-script](https://github.com/BrianHicks/nix-script)
+  - Nix version of scriptisto, letting you use any Nix dependency
+
+See also [Single-file scripts that download their dependencies](https://dbohdan.com/scripts-with-dependencies)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Wherther `version` should may be unset or defaulted to `0.0.0`
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Note: we are assuming the following are **not** future possibilities in this design
+- Embedding build scripts
+- Embedding `.cargo/config.toml` files
+- Embedding `rust-toolchain.toml` files
+- Embedding other source files or additional packages
+
+## Playground support
+
+The [playground](play.rust-lang.org/) could gain support for embedded
+manifests, allowing users access to more packages, specific package versions,
+and specific feature sets.
+
+## Dealing with leaked `target/`
+
+As `target/` is out of sight, it is easy to "leak" them, eating up disk space.
+Users would need to know to run `cargo clean --manifest-path foo.rs` before
+deleting `foo.rs` or to just do `rm -rf ~/.cargo/target` and wipe everything
+(since the specific target directory will be non-obvious).
+
+In the future we can
+[track embedded manifests and garbage collect their `target/`](https://github.com/rust-lang/cargo/issues/12633).
+
+## `cargo new` support
+
+A `cargo new <flag> foo.rs` could generate a source file with an embedded manifest.
+
+Questions
+- Should we do this implicitly with the `.rs` extension?
+- If we have a flag, what should we name it?
```

Comment:
I'd go with `--script`, though I'm not sure if `cargo new` really needs to support it if the goal is to be as boilerplate-free as possible